### PR TITLE
DB subnet migration

### PR DESCRIPTION
**Pull request by amarin-dspace**
- removed separate, private DB subnets and (re-)used already existing VPC private subnets
- cleaned up unnecessary parts of code
- added tags to VPC private and public subnets, to more easily differentiate between them
- introduced CHANGELOG.md file, to describe changes between tagged releases
- introduced MIGRATION.md file, to describe migration steps between tagged releases (if introducing breaking changes)
- tested on clean deployment and already existing deployment
- migration steps described in MIGRATION.md file
- verified to be working

Terraform v1.7.1
on linux_amd64
+ provider registry.terraform.io/gavinbunney/kubectl v1.14.0
+ provider registry.terraform.io/hashicorp/aws v4.67.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.3.3
+ provider registry.terraform.io/hashicorp/helm v2.12.1
+ provider registry.terraform.io/hashicorp/http v3.4.1
+ provider registry.terraform.io/hashicorp/kubernetes v2.26.0
+ provider registry.terraform.io/hashicorp/local v2.4.1
+ provider registry.terraform.io/hashicorp/null v3.2.2
+ provider registry.terraform.io/hashicorp/random v3.6.0
+ provider registry.terraform.io/hashicorp/time v0.10.0
+ provider registry.terraform.io/hashicorp/tls v4.0.5
+ provider registry.terraform.io/terraform-aws-modules/http v2.4.1

### CODE REVIEW

File: modules/simphera_aws_instance/postgresql.tf [hunk 1]
```diff
@@ -1,3 +1,17 @@
+data "aws_subnet" "example" {
```
Comment (mariogluhakovic) [inline modules/simphera_aws_instance/postgresql.tf:1]: This example should be changed with something meaningful like vpc_private_subnets, private_subnets or something like that and then reuse the same name in line 8.

File: modules/simphera_aws_instance/postgresql.tf [hunk 2]
```diff
@@ -1,3 +1,17 @@
+data "aws_subnet" "example" {
+  for_each = toset(var.private_subnets)
+  id       = each.value
+}
+
+resource "aws_db_subnet_group" "database" {
```
Comment (mariogluhakovic) [inline modules/simphera_aws_instance/postgresql.tf:6]: I recommend to change naming to something like database_subnet_group or something similar since resource is not database but subnet group of database.

**Comment by schwichti** [comment]
Why the DB subnets have to be replaced?

File: modules/simphera_aws_instance/postgresql.tf (hunk 1 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf]: Done

File: modules/simphera_aws_instance/postgresql.tf (hunk 2 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf:6]: This is aligned with previous setup (same naming etc). You can check the steps in MIGRATION.md and see the previous names of resources.

**Comment by schwichti** [comment]
> @schwichti The aim was simplification, ie to have as less of unnecessary configurations as possible, plus it's easier to read and maintain the code (especially in this case, where some configuration options are hidden/implicit, variable values are wrong and/or implicit etc). Functionally, nothing changes, RDS is still on private subnets, is not publicly exposed.

What are the advantages of DB subnets? What is the reason why AWS offers specialized DB subnets?

**Comment by amarin-dspace** [comment]
> What are the advantages of DB subnets? What is the reason why AWS offers specialized DB subnets?

There are no "specialized" DB subnets, just DB subnet group. Subnets are part of VPC, there are some advantages in having network separation between components, but it doesn't really matter here. So, with this PR, we're simply reusing existing private subnets, which we create together with VPC, instead of creating 3 new private subnets and populating DB subnet group with them. While we incur no additional costs, and we're not even close to subnet quota, it's still wasteful to have /22 subnets and use only couple of IPs from there (on top of other benefits, already mentioned).

File: modules/simphera_aws_instance/postgresql.tf [hunk 3]
```diff
@@ -1,3 +1,17 @@
+data "aws_subnet" "vpc_private_subnet" {
+  for_each = toset(var.private_subnets)
+  id       = each.value
+}
+
+resource "aws_db_subnet_group" "database" {
+  name       = "${var.infrastructurename}-vpc"
+  subnet_ids = [for s in data.aws_subnet.vpc_private_subnet : s.id]
```
Comment (schwichti) [inline modules/simphera_aws_instance/postgresql.tf:8]: Can you use `module.vpc.private_subnets`? In that case you would not need `data.aws_subnet`

File: MIGRATION.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#Migrate AWS RDS databases to use different private subnets
+
+This procedure explains how to migrate AWS RDS databases, deployed with ref architecture, from using separate, private DB subnets to (re-)using 
+private subnets already deployed with VPC. This only applies to already existing deployments, new deployments are not affected.
+The change was introduced with v0.2.0 of reference architecture.
+
+##Steps:
+1. Check your existing deployment and make sure everything works etc. (no issues, pods not starting and so on)
+2. Delete keycloak and simphera databases, making sure that final snapshot is created
```
Comment (schwichti) [inline MIGRATION.md:9]: Can you provide more details to create a snapshot.

File: MIGRATION.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#Migrate AWS RDS databases to use different private subnets
+
+This procedure explains how to migrate AWS RDS databases, deployed with ref architecture, from using separate, private DB subnets to (re-)using
```
Comment (schwichti) [inline MIGRATION.md:3]: ref architecture => reference architecture

File: MIGRATION.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+This procedure explains how to migrate AWS RDS databases, deployed with ref architecture, from using separate, private DB subnets to (re-)using 
+private subnets already deployed with VPC. This only applies to already existing deployments, new deployments are not affected.
+The change was introduced with v0.2.0 of reference architecture.
+
+##Steps:
+1. Check your existing deployment and make sure everything works etc. (no issues, pods not starting and so on)
+2. Delete keycloak and simphera databases, making sure that final snapshot is created
+3. After DBs are deleted, update DB subnet group, remove DB specific subnets and add VPC private subnets
+4. Restore DBs from snapshots taken, selecting same configuration options as original DBs (names, VPCs, security groups etc.)
+5. Remove old DB subnet group from TF state, eg.: terraform state rm module.vpc.aws_db_subnet_group.database[0]
```
Comment (schwichti) [inline MIGRATION.md:12]: TF => Terraform (all occurences)

File: MIGRATION.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+6. Import new DB subnet group into TF state, eg.: terraform import module.simphera_instance[\"production\"].aws_db_subnet_group.default <infrastructure name>-vpc
+7. Re-run TF to refresh the state and apply any small configuration changes in-place, which were missed in previous steps
```
Comment (schwichti) [inline MIGRATION.md:14]: Suggestion:
7. Let Terraform apply the remaining changes and refresh the Terraform state by re-running `terraform apply`

**Review by schwichti — CHANGES_REQUESTED** [review]
Small comments

File: MIGRATION.md (hunk 4 above)
Comment (amarin-dspace) [reply MIGRATION.md:9]: It's a default option (automatically selected) when deleting a DB, no special input is needed.

File: MIGRATION.md (hunk 7 above)
Comment (amarin-dspace) [reply MIGRATION.md]: Done

File: MIGRATION.md (hunk 6 above)
Comment (amarin-dspace) [reply MIGRATION.md]: Done

File: MIGRATION.md (hunk 5 above)
Comment (amarin-dspace) [reply MIGRATION.md]: Done

File: modules/simphera_aws_instance/postgresql.tf (hunk 3 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf:8]: Yes, a good catch, it will not work with clean deployment now, when filter is replaced with var. Let me fix that.

File: MIGRATION.md (hunk 4 above)
Comment (schwichti) [reply MIGRATION.md]: You could add a hint like "When deleting the database via the AWS Console, the option to create a snapshot is offered."

Comment (amarin-dspace) [reply MIGRATION.md]: Done

File: modules/simphera_aws_instance/postgresql.tf (hunk 3 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf]: Fixed

File: modules/simphera_aws_instance/postgresql.tf [hunk 8]
```diff
@@ -1,3 +1,17 @@
+data "aws_subnet" "vpc_private_subnet" {
```
Comment (schwichti) [inline modules/simphera_aws_instance/postgresql.tf:1]: Why is this necessary? Why can't we use `var.private_subnets` directly?

Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf]: Yeah, I guess we can use it directly; https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws/latest#output_private_subnets
Checking value passing in other ways is more interesting :-)

File: modules/simphera_aws_instance/postgresql.tf [hunk 9]
```diff
@@ -1,3 +1,12 @@
+resource "aws_db_subnet_group" "database" {
+  name       = "${var.infrastructurename}-vpc"
+  subnet_ids = var.private_subnets
+
+  tags = {
+    Name = "My DB subnet group"
```
Comment (schwichti) [inline modules/simphera_aws_instance/postgresql.tf:6]: please delete this tag.

File: modules/simphera_aws_instance/postgresql.tf [hunk 10]
```diff
@@ -1,3 +1,12 @@
+resource "aws_db_subnet_group" "database" {
+  name       = "${var.infrastructurename}-vpc"
```
Comment (schwichti) [inline modules/simphera_aws_instance/postgresql.tf:2]: Very important: Use `name = "${local.instancename}-vpc"` Otherwise you will get a conflict when you have multiple simphera instances defined in `var.simpheraInstances`

Comment (schwichti) [inline modules/simphera_aws_instance/postgresql.tf:2]: Suggestion: Use a more self-descriptive suffix like `-dbsubnet` instead of `-vpc`

**Review by schwichti — CHANGES_REQUESTED** [review]

File: modules/simphera_aws_instance/postgresql.tf (hunk 9 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf:6]: Well, we should have tags on our resources, it's a good practice and easier to track resources. I expected a comment on this, but more towards value (ie "what should we put as tag") rather than "do we need a tag here"?

File: modules/simphera_aws_instance/postgresql.tf (hunk 10 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf:2]: Hm, ok, makes sense, but not sure of any case where multiple simphera instances are defined. In any case, will update this.

Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf:2]: This is done to align with previous naming convention (before db subnet migration) and is aligned with migration doc. If you follow the steps, there will be no re-deployment of Simphera resources, only possibly update in-place.

Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf:2]: Done

Comment (schwichti) [reply modules/simphera_aws_instance/postgresql.tf]: Indeed, the feature is currently barely used. The idea behind it is to have different SIMPHERA stages that share the same EKS etc. It was adopted from the Azure reference architecuture. I expect the feature to become more important in the future.

File: modules/simphera_aws_instance/postgresql.tf (hunk 9 above)
Comment (amarin-dspace) [reply modules/simphera_aws_instance/postgresql.tf]: Done

**Review by schwichti — APPROVED** [review]
